src/pages/Chat/Chat/index.tsx

```diff
@@ -0,0 +1,326 @@
+  const firstLoad = useRef<boolean>(true);
+  const [dialog, setDialog] = useState<boolean>(false);
+
+  const sorter = (a: IMessage, b: IMessage): number => (b.createdAt as Date).getTime() - (a.createdAt as Date).getTime();
+
+  const naiveUnique = (msgs: IMessage[]): IMessage[] => {
+    // eslint-disable-next-line no-underscore-dangle
+    const messagesId = msgs.map((m) => m._id);
+
+    // eslint-disable-next-line no-underscore-dangle
+    return msgs.filter((m, i) => messagesId.indexOf(m._id) === i);
+  };
+
+  const addMessages = useCallback((loadedMessages: QuerySnapshotDocuments): void => {
+    const chatMessages = loadedMessages.map((message) => {
+      const messageData = message.data();
+      return {
+        _id: message.id,
+        text: messageData.text,
+        createdAt: messageData.timestamp ? messageData.timestamp.toDate() : new Date(),
+        user: {
+          _id: messageData.sender.id,
+        },
+      };
+    });
+
+    if (chatMessages.length === 0) return;
+
+    setMessages((previousMessages) => naiveUnique(GiftedChat.append(previousMessages, chatMessages).sort(sorter)));
+  }, []);
+
+  useEffect((): void => {
+    if (firstLoad.current || !chatRef) {
+      return;
+    }
+
+    // eslint-disable-next-line no-underscore-dangle
+    let newestDate = messages.filter((m) => m.user._id !== currentUserRef.id)[0]?.createdAt as Date;
+    // If there's no other user message yet, use newest from current user
+    if (!newestDate) {
+      newestDate = messages[0]?.createdAt as Date;
+    }
+
+    if (newestDate) {
+      chatAPI.loadNewMessages(chatRef, newestDate).then((loadedMessages) => {
+        const filteredLoadedMessages = loadedMessages.docs.filter((doc) => doc.data().sender.id !== currentUserRef.id);
+        addMessages(filteredLoadedMessages);
+      });
+    }
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [shouldUpdate]);
+
+  const loadMessages = useCallback((lastMessage?: IMessage): void => {
+    if (!chatRef) return;
+
+    const lastDate = lastMessage ? lastMessage.createdAt as Date : undefined;
+
+    chatAPI.loadMessages(chatRef, 10, lastDate)
+      .then((loadedMessages) => {
+        addMessages(loadedMessages.docs);
+        setIsLoadingEarlier(false);
+        firstLoad.current = false;
+      })
+      .catch(() => {
+        Alert.alert(
+          'Erro!',
+          'Ocorreu um erro no carregamento das informações.\n'
+              + 'Retornando para a página anterior.',
+          [
+            {
+              text: 'Ok',
+              onPress: () => navigation.goBack(),
+            },
+          ],
+          {
+            cancelable: false,
+          },
+        );
+      });
+  }, [chatRef, navigation, addMessages]);
+
+  const updateTyping = (): void => {
+    // onInputTextChanged triggers when first load
+    if (ignoreFirstOnInputTextChanged.current === true) {
+      ignoreFirstOnInputTextChanged.current = false;
```

Comment:
If we are going to set this variable to `false`, maybe we should just call it `firstOnInputTextChanged` or `ignoreOnInputTextChanged` or some other name. That way we convey the idea that we ignore the first input only better.